### PR TITLE
Add support for nested metric values. (#959)

### PR DESCRIPTION
**Pull request by isolet**
Description:

This commit add support for nested metric values. If `Metric.compute` returns a `dict`, its values will be flattened into `engine.state.metrics` in `Metric.completed`. 

Note that, in this case the value types of the returned `dict` is limited(i.e. `Dict[Any, non number type]` is not allowed), while the original allowed type of `Metric.compute` is `Any`. We can also allow other types of `dict` values, but this may be lead unclear behaviors to users when they try to return (different kinds of )`dict`s.

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/metric.py [hunk 1]
```diff
@@ -114,6 +116,19 @@ def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union
             return tensor.item()
         return tensor
 
+    def ensure_metric_value(self, value: Any, raise_error=True) -> Union[numbers.Number, torch.Tensor]:
```
Comment (sdesrozis) [inline ignite/metrics/metric.py:119]: I don’t know if an exception should be raised. Metrics are very generics, it can work with ˋAny`. Metrics associated to loggers should use restricted types. 

@vfdev-5 ?

Comment (isolet) [reply ignite/metrics/metric.py:119]: I agree metrics are generics. Maybe remove type checking here and move to `Logger` later when necessary?

Comment (sdesrozis) [reply ignite/metrics/metric.py:119]: Ok, let’s try this!

Comment (isolet) [reply ignite/metrics/metric.py:119]: I just found one more issue, if `Any` type is allowed, shall we deep flatten a dict in this case? Or shall me only care about `dict` contains all number values?

Comment (vfdev-5) [reply ignite/metrics/metric.py:119]: IMO, we should not check the output type and raise an error and just keep previous behaviour.

File: ignite/metrics/metric.py [hunk 2]
```diff
@@ -137,9 +152,12 @@ def iteration_completed(self, engine: Engine) -> None:
 
     def completed(self, engine: Engine, name: str) -> None:
         result = self.compute()
-        if torch.is_tensor(result) and len(result.shape) == 0:
-            result = result.item()
-        engine.state.metrics[name] = result
+        if isinstance(result, dict):
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:155]:
```suggestion
        if isinstance(result, Mapping):
```

**Comment by vfdev-5** [comment]
IMO, idea of the PR is to unnest one level of mappings (e.g. dict) as @isolet's use-case. Thoughts ?

File: ignite/metrics/metric.py (hunk 1 above)
Comment (sdesrozis) [reply ignite/metrics/metric.py:119]: I agree. At the moment, we should flat only the result of metrics. IMHO, what is embedded in dict is related to the metric itself.

We would notice in doc that generic result type ˋAny` is a generic purpose but interoperability with others handlers / features implies some restrictions : scalars (number, tensor) and dict of scalars.

What do you think ?

**Comment by isolet** [comment]
My original thought is single enough just to flatten a level one dict of metrics. But as metrics is so generic, now I think maybe we should also not to make level one dict anything special? Before I opened the issue, I used the following snippet to get my goal.

```py


class MultipleMetricsOutputHandler(OutputHandler):

  def _setup_output_metrics(self, engine):
      metrics = {}
      if self.metric_names is not None:
          if (isinstance(self.metric_names, str)
                  and self.metric_names == 'all'):
              metrics = flatten_dict(engine.state.metrics)
          else:
              for name in self.metric_names:
                  if name not in engine.state.metrics:
                      warnings.warn('Provided metric name \'{}\' is missing '
                                    'in engine\'s state metrics: {}'.format(
                                        name,
                                        list(engine.state.metrics.keys())))
                      continue
                  metrics[name] = engine.state.metrics[name]

      if self.output_transform is not None:
          output_dict = self.output_transform(engine.state.output)

          if not isinstance(output_dict, dict):
              output_dict = {'output': output_dict}

          metrics.update(
              {name: value
               for name, value in output_dict.items()})

      return metrics
```

How about keeping anything of `Metric` unchanged? Is there a better place to achieve my goal? I'm not familiar enough with `ignite` so I can not get an answer now ...

**Comment by vfdev-5** [comment]
@isolet I think one level flattening is a good level of simplifying things and keeping Metric rather generic. For other non-trivial cases, use should write its adapters.

File: ignite/metrics/metric.py [hunk 3]
```diff
@@ -82,7 +82,9 @@ def compute(self) -> Any:
         This is called at the end of each epoch.
 
         Returns:
-            Any: the actual quantity of interest.
+            Any: the actual quantity of interest. However, if a :class:`dict` is returned, it should contain
+                 :class:`numbers.Number` or :class:`torch.tensor` values and will flattened into
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:86]: Let's remove those type limitations:
```
Any: the actual quantity of interest. However, if a :class:`dict` is returned, it will flattened into  `engine.state.metrics` when :func:`~ignite.metrics.Metric.completed` is called.
```

Comment (isolet) [reply ignite/metrics/metric.py]: Done!

**Comment by sdesrozis** [comment]
LGTM but CI quite unstable...

**Comment by isolet** [comment]
Thanks!
